packages/remix-dev/compiler.ts

```diff
@@ -442,6 +442,12 @@ function createServerBuild(
       format: config.serverModuleFormat,
       treeShaking: true,
       minify: options.mode === BuildMode.Production && isCloudflareRuntime,
+      // The type of dead code elimination we want to do depends on the
+      // minify syntax property: https://github.com/evanw/esbuild/issues/672#issuecomment-1029682369
+      // without this we have dev builds that run fine in production, but
+      // ship server code to the browser, blowing it up due to not optimizing
+      // out imports and module level blocks like `if (process.env.NODE_ENV === "test") {`.
+      minifySyntax: true,
```

Comment:
Total nitpick, but I might suggest putting `minifySyntax: true` right below `treeShaking: true` since both are targeting the same end goal (removing code) and both are non-conditional.  Then `minify` comes third and builds on top of `minifySyntax`